docs/src/nondiff_points.md

```diff
@@ -0,0 +1,155 @@
+For a function like `abs` or `x -> x<0 ? x : -2x` setting the non-differentiable point to either side would result in it leaving that maxima.
+
+Further, a nice (albeit generally intractable) algorithm for finding the global optima is to take the set of all stationary points (i.e. points where the derivative is zero), combine that with the boundary points and evaluate the primal function at all of them.
+The extrema of this set are the global optima.
+This algorithm is only guaranteed correct for functions that are differentiable everywhere *or* that apply the sub/super-derivative convention and make all non-differentiable local optima claim to have a zero derivative.
+It is also correct if you make other non-differentiable points have zero derivative, just slower.
+
+### It is sometimes the non-zero, especially if it boarders a flat section
+
+If a function has derivative zero in some large section of it's domain, like `relu` or `x->clamp(x, -1, 1)`,
+a case can be made for choosing the non-zero derivative branch.
+Depending exactly on the larger function being optimized, it is often the case that a zero gradient for this function means a total zero gradient (e.g. if the greater function is a chain of composed calls).
+So once things move into the flat-region they stay there.
+If we chose the other branch we move them into (just barely) the nonflat region.
+If moving into the non-flat region was good they will move further there later.
+If it is not good then they well be rapidly cast deeper into the flat region and we will not be at this boundary case.
+
+### It is sometimes the mean
+Either the mean of the branches, or the overall mean of the sub/super-differential (which may be equal).
+
+A nice advantage of it is that it will agree with the result you will get from central finite differencing.
+Which is what [ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl) defaults to.
+
+### All else being equal it is just one of the branches
+Pick one.
+It will be fast.
+It will agree with either the forwards or reverse finite differencing results.
+
+### It is almost never `Inf`, `NaN` or an error
+Practically speaking, people are generally using AD to perform some gradient descent like optimization procedure.
+`NaN` and `Inf` do not generally take us to nice places.
+Erroring can be worse, especially if it is a local minima -- the optimization fully converges to that minima and then throws an error rather than reporting the result.
+If the primal function takes a non-finite value or errors on one side, then we are in the case that we are at a boundary of the domain.
+Which means we are free to chose *any* value.
+In particular we often want to chose value of the derivative from **the other side where the primal is defined.**
+
+
+## Why is the sub/super-differential convention permissible?
+
+Math isn't real, and AD doesn't do calculus.
+We are trying to accomplish some goal, and this approach works well.
```

Comment:
So when I read this I feel it claims we're doing something inconsistent in the name of pragmatism.

This is funny, but I feel like it obscures something deeper going on here. Now, I'm no AD expert... but is the story something like:

* AD is a machine for applying the chain rule (forget deriving it from calculus; consider it purely as an algebraic rule in its own right)
* AD relies on a root set of rules for elementary functions; the (function,rule) pairs can be treated as their own objects
* Given the (function,rule) definitions, AD coherently and consistently applies the chain rule; it's all just algebraic at this point.

We didn't say anything about calculus yet.

* To do calculus, we define rules which *happen* to be consistent with the normal limits of calculus, where they apply.
* With this definition, AD does compute derivatives over most parts of the space, and does a part of the thing we call calculus.
* For cases where normal limits have nothing to say about the derivative, we may simply choose something. This choice doesn't make our algebraic interpretation of the AD machine *inconsistent*, though the free choice does make the result somewhat *arbitrary*.
* If we choose a sub/super-derivative there's a good chance the result after the chain rule will also be one (though as I see you discussed above, not guaranteed).

With this point of view in mind, it's unclear to me why it's even necessary to choose a sub/super-differential for consistency. Here's where the pragmatism comes in - we want to do optimization, and getting  a sub/super-differential out after applying the chain rule is less likely to lead to practical problems than getting some completely nonsense value!

Sometimes there's freedom of choice in math. This isn't necessarily a sign of inconsistency!

Reply:
right yes. It is a flippant comment that can probably be removed

Reply:
Actually I found the comment amusing :-)

But more importantly, do you think this assessment of what's going on is accurate?